src/core/react/ColorPicker.tsx

```diff
@@ -56,6 +57,20 @@ export const ColorPicker = React.memo((props: ColorPickerProps) => {
     const [colorHex, setColorHex] = React.useState(color.toString());
     const [visible, setVisible] = React.useState<boolean>(false);
 
+    const colorsCurrentlyUsed = useSelector((state) => Object.values(state['editor']['presentState']['state']['values']['diagrams']['items']));
+
+const colorsCurrentlyUsedInEditor = (theArray: any) => {
+     let result: any = [];
+     theArray.forEach((element: any) => {
+     for (let key in element.values.items.items) {
```

Comment:
The colors need to be calculated outside so that we can also serialize them properly. Perhaps in the EditorState and then with a reducer.

Reply:
Can you please provide more details here? The idea is to use a reducer to calculate outside the 'ColorPicker' component, correct?  Is the correct approach to add a new Key containing the values of the editor colors? (this would be created under 'presentState' > 'state' > 'values')

Also what do you mean why serializing the colors properly?

Reply:
First of all this calculation should not be part of the color picker. There are several reasons:

1. Circular dependencies, because the color picker is part of a lower, utility layer and should not reference other layers.
2. Reusability, it makes the color picker harder to use for other elements.
3. Performance: There are several color pickers and you would have to make the calculation several times.
4. Testing: There is a story for that where we should also test the recent colors and it would make testing much harder.

So if we move it to another layer there are 2 opportunities:

## 1. Write a selector

Write a selector with 'reselect': https://github.com/mydraft-cc/ui/blob/4fd63c7005b7f9963830ac7757208ba2efb8fafd/src/wireframes/model/projections.ts. The big advantage is that the result is cached and we only calculate it once. We have to test performance though. If we have 10 pages with 200 elements each and each element has like 10 appearance items we would have to loop over 10 * 200 * 10 = 20000 elements each time, something is changed. The advantage is that the implementation is easier and straight forward.

I would test this approach first and make a performance test, you could also write a unit test for that, where you measure it.

## 2. Add the colors to the store

The state is hold in classes: https://github.com/mydraft-cc/ui/blob/master/src/wireframes/model/editor-state.ts, we could also store the colors there. The advantage is, that we can keep the colors there and only calculate them when it is needed:

1. An item is removed
2. An item is added
3. An item is added with the copy & paste.
4. An item appearance has been changed.

The "problem" is that we are updating the list of colors, we need a object like `{ [color: string]: number }` and store how often the color is used, so that we can also remove colors when they are not referenced anymore.

> Also what do you mean why serializing the colors properly?

If we store the colors in the editor state, we also need to serialize them, so that they become part of the file that is being saved to disk or to server. You can find the code here: https://github.com/mydraft-cc/ui/blob/master/src/wireframes/model/serializer.ts

Reply:
> 2. Add the colors to the store
I'm reviewing how this store works since it doesn't appear to be Redux. Is there documentation available on the basic structure? Trying to see how I can view the EditorState. I can only see one stored icon, even though I have multiple saved.

Reply:
It is 100% in redux. There is no documentation, but it is basically like this:

* UndoRedoState (maintains multiple copies of the state for undo redo):
   * present: EditorState (Current state root)
       * diagrams (1...N) multiple diagrams (aka page)
           * items (1....N) multiple items like shapes or groups
              * visuals (the properties like background, foreground and so on)
              * id
              * transform (size + position)
              * type (whether it is a group or a shape)

Reply:
In my new commit, I added a Selector to retrieve the document colors. However, I'm seeing a "Circular dependency detected" error, I'm currently debugging it, but wanted to reach out for guidance on how to correct this issue.